### ISSUE TITLE
Dask Log Server

### ISSUE THREAD

**Issue by JulianWgs** [dask/dask#6367]
Hello Dask Team,

I've created a small tool which saves all the task graphs and tasks information (duration etc.), versions and system information from a client during execution. My plan is to develop this into a tool similiar to the Spark history server.

> The goal of this project is to preserve all necessary runtime data of a Dask client in order to "replay" and analyze the performance and behavior of the client after the fact.

Have a look at the project site:
https://github.com/JulianWgs/dask-log-server

I've also a few questions:
- Which logging features are already in place in dask? For example there are loggers in every module ([here](https://github.com/dask/dask-ml/blob/4fe3af2d26d7d83a673befd40b32eab049258a75/dask_ml/cluster/k_means.py#L157)). I couldn't find an overview. Is this information worth logging?
- Is something like this needed?
- How would you go about integrating this into the dashboard, since it's only for real time data?
- What would a scalable logging format look like? Right now I'm using jsonl.

With best regards,
Julian

**Comment by TomAugspurger** [comment]
> I couldn't find an overview. Is this information worth logging?

Yes, that seems reasonable.

Just want to make sure you're also aware of Dask's performance report: https://distributed.dask.org/en/latest/diagnosing-performance.html#performance-reports and `get_task_stream`: https://distributed.dask.org/en/latest/api.html#distributed.Client.get_task_stream

**Comment by JulianWgs** [comment]
Thanks for your response.

The problem I see with the performance reports are twofold. Firstly they have to be created purposely. I can't decide to create them for computions of the past. At least I'm not aware of that option. I want a soluation which just runs in the background and where I can produce these performance reports afterwards. That's actually sounds like another usecase I want to incorporate. Thanks for the suggestion! Secondly they don't offer you the raw data for further investigation (again at least I'm not aware of it) or to reproduce the graphs/visualization for a paper.

I'm logging the task stream every x seconds if they have been more then y tasks submitted to the client.

With best regards,
Julian

**Comment by mrocklin** [comment]
I think that a server makes sense.  The task stream does actually give you
the raw data (see the .data attribute) which was designed for exactly this
purpose.  The fact that you have to opt-in to it does make it less
useful or information gathering though.

**Comment by JulianWgs** [comment]
I've added a dask task stream enhanced version of the dask.visualize function. You can set the node labels or the fill color (both numerical and categorical values are supported) to a value in task stream dataframe. Since basically any data can be joined to that dataframe possibilities are endless.

Here is an example of a small task graph with fill color set to the worker it was executed on and the label to the start date. See the doc string of the function for an explanation of all the features.

![image](https://user-images.githubusercontent.com/31596773/86538778-6a21fa80-bef8-11ea-8a54-537a25aad222.png)

Unfortunately I've ran into an issue. If the task graph is only built once, but executed multiple time it is not possible to assign each row in the task stream dataframe to one specific execution. One option would be to sort the dataframe by start and then dividing it into the executions. However two graphs could be executed at the same time or data could be missing. This would lead to errors. Would it be possible to add a unique hash to the task stream to identify each execution from one another?

Code would look like this:

```python
from dask import delayed
import dask

@delayed
def add(a, b):
    return a + b

@delayed
def inc(a):
    return a + 1

dask_graph = add(inc(2), inc(3))
dask.compute(dask_graph)
dask.compute(dask_graph)
```

Thank you!

With best regards
Julian

**Comment by TomAugspurger** [comment]
> Would it be possible to add a unique hash to the task stream to identify each execution from one another?

I don't really understand what this would look like. But IIUC, it sounds like it doesn't fit well with the rest of Dask.

**Comment by JulianWgs** [comment]
Why do you think it does not fit well? I don't think it would need a large change or would it?

I've two ideas on how to implement this:
- add another hash to the key of the task before the execution. This way it is unique even when executed multiple time.
- add an id column to the task stream to identify each execution of a graph.

This is how the task stream data looks like right now:
![grafik](https://user-images.githubusercontent.com/31596773/86602072-6fd41a80-bfa2-11ea-9323-83206214f92d.png)

I want map each task to their graph. That is easily done via the key column, when the same graph is only executed once, because each key is unique. However when the same graph is executed muliple times it is hard to do the mapping this way.

**Comment by JulianWgs** [comment]
Hello,

I just wanted to give you an upate

- I've added a function to export the dask graph to networkx which might be interesting for all kinds of graph analysis tasks. I've yet added a function which determines the maximum possible concurrency of a graph, but there is more to come.
- I've created a demo website [here](https://dask-log-server.herokuapp.com/). Unfortunately I can't install graphviz on Heroku, so the graph display does not work.
- I've completed to implement all necessary loggers, which includes configuration, versions, task stream, info (CPU usage, etc.) and additional loggers by name. Am I missing anything?

I'm still in need of a solution of the above mentioned problem. If multiple tasks with the same key are executed it is impossible to assign them to a specific graph / run.

Best regards,
Julian
